src/lib/testResultGenerator.ts

```diff
@@ -0,0 +1,323 @@
+   testPlanFilename: string,
+   reportFilename: string,
+   format: Format
+): Promise<TestReport> {
+   const testPlan: TestPlan = await loadTestPlanFromFile(testPlanFilename);
+
+   const featureResults: FeatureResults[] = await loadTestResults(
+      featureRootDirectory
+   );
+
+   const testResults: TestResult[] = evaluateTestResults(
+      testPlan.versionToTest,
+      testPlan.testCases,
+      featureResults
+   );
+
+   const testReport = {
+      versionToTest: testPlan.versionToTest,
+      testResults: testResults,
+   };
+
+   exportTo(format, testReport, reportFilename);
+   return testReport;
+}
+
+/**
+ * Gather all the test results from the feature files
+ * @param directory
+ */
+export async function loadTestResults(
+   directory: string
+): Promise<FeatureResults[]> {
+   const features: Feature[] = await featureProcessor.loadFeaturesFrom(
+      directory
+   );
+
+   return features.map(feature => {
+      const scenarioResults = getScenarioResults(feature);
+      const version = getFeatureVersion(scenarioResults);
+
+      return {
+         name: feature.filename,
+         version: version,
+         scenarios: scenarioResults,
+      };
+   });
+}
+
+/**
+ * Gather test results for this feature's scenarios
+ * @param feature
+ */
+function getScenarioResults(feature: Feature): ScenarioResults[] {
+   return (
+      sortBy(feature.children, "scenario.name")
+         // keep only valid scenarios
+         .filter(child => !!child && !!child.scenario)
+         .map(child => child.scenario)
+         .map(scenario => {
+            const name = scenario.name;
+            const stepResults = getStepResults(scenario);
+            const version = getScenarioVersion(stepResults);
+            return {
+               name: name,
+               version: version,
+               steps: stepResults,
+            };
+         })
+   );
+}
+
+/**
+ * Gather test results for this scenarios's steps
+ * which is located in a cucumber table row
+ * @param scenario
+ */
+function getStepResults(scenario: IScenario): StepResults[] {
+   return scenario.steps
+      .filter(
+         step =>
+            !!step &&
+            !!step.dataTable &&
+            !!step.dataTable.rows &&
+            step.dataTable.rows.length > 0
+      )
+      .map(step => getLatestStepResult(step));
+}
+
+/**
+ * Find latest result from the step history
+ * FORMAT:
+ *    | test step status | version tested | tester Name | defect number  |
+ *    example
+ *    | FAIL | 0.39.00-v201811141002 | Tester One | defect-1 |
+ * @param step
+ */
+function getLatestStepResult(step: IStep): StepResults {
+   const rows = step.dataTable.rows;
+
+   // Keep only the latest test result
+   const reducer = (latest: string, row: ITableRow): string => {
+      const stepVersion = !!row.cells
+         ? row.cells[1].value.toString().trim()
+         : undefined;
+      return !!stepVersion && stepVersion > latest ? stepVersion : latest;
```

Comment:
I don't understand this one.  The version will be something like "0.40.201901011100"  How does this cast to number?

Reply:
The concern here is that the string sort is not a natural sort.  Depending on the version in the test plan, version 40.0.1 (which is minor version) might sort lower than 40.0.20190125 (which is a date).  

However, because version strings can vary in so many ways, the assumption that is the test plan  will use a string that sorts adequately.  

For example, version to test: 40.0.20190125 (January 25th 2019)
versions in the feature file: 0.39.201901 - 0.40.0.20190129 (January 29th 2019)